hera_pspec/grouping.py

```diff
@@ -496,13 +511,312 @@ def average_spectra(uvp_in, blpair_groups=None, time_avg=False,
+        data_array[spw] = np.zeros((uvp.Ntimes, Ndlyblps, uvp.Npols), dtype=np.complex128)
+        if store_cov:
+            cov_array[spw] = np.zeros((uvp.Ntimes, Ndlyblps, Ndlyblps, uvp.Npols), dtype=np.complex128)
+        if store_stats:
+            for stat in uvp.stats_array.keys():
+                stats_array[stat][spw] = np.zeros((uvp.Ntimes, Ndlyblps, uvp.Npols), dtype=np.complex128)
+        if store_window:
+            window_function_array[spw] = np.zeros((uvp.Ntimes, Ndlyblps, Ndlyblps, uvp.Npols), dtype=np.complex128)
+ 
+        # setup the design matrix: P_cyl = A P_sph
+        A[spw] = np.zeros((uvp.Ntimes, Ndlyblps, Nk, uvp.Npols), dtype=np.float64)
+
+        # setup weighting matrix: block diagonal for each Ndly x Ndly
+        E = np.zeros((uvp.Ntimes, Ndlyblps, Ndlyblps, uvp.Npols), dtype=np.float64)
+
+        # get kperps for this spw: shape (Nblpairts,)
+        kperps = uvp.get_kperps(spw, little_h=True)
+
+        # get kparas for this spw: shape (Ndlys,)
+        kparas = uvp.get_kparas(spw, little_h=True)
+
+        # get k to tau mapping for this spw
+        avgz = uvp.cosmo.f2z(np.mean(uvp.freq_array[uvp.spw_to_freq_indices(spw)]))
+        t2k = uvp.cosmo.tau_to_kpara(avgz, little_h=True)
+        taus = kbins / t2k
+        dlys_array.extend(taus)
+
+        # store kbins as delay bins
+        spw_dlys_array.extend(np.ones_like(taus, dtype=np.int) * spw)
+
+        # iterate over blpairs
+        for b, blp in enumerate(uvp.get_blpairs()):
+            # get blpair time indices
+            blpt_inds = uvp.blpair_to_indices(blp)
+
+            # get k magnitude of data: (Ndlys,)
+            kmags = np.sqrt(kperps[blpt_inds][0]**2 + kparas**2)
+
+            # shape of nsmap: (Ntimes, Npols)
+            nsmp = uvp.nsample_array[spw][blpt_inds]
+
+            # shape of wgts: (Ntimes, spw_Nfreqs, 2, Npols)
+            wgts = uvp.wgt_array[spw][blpt_inds]
+
+            # shape of ints: (Ntimes, Npols)
+            ints = uvp.integration_array[spw][blpt_inds]
+
+            # get Ndlyblps slice
+            dstart, dstop = Ndlys * b, Ndlys * (b + 1)
+            dslice = slice(dstart, dstop)
+
+            # store data
+            data_array[spw][:, dslice] = uvp.data_array[spw][blpt_inds]
+            if store_window:
+                window_function_array[spw][:, dslice, dslice] = uvp.window_function_array[spw][blpt_inds]
+            if store_stats:
+                for stat in stats_array:
+                    stats_array[stat][spw][:, dslice] = uvp.stats_array[stat][spw][blpt_inds]
+            if store_cov:
+                cov_array[spw][:, dslice, dslice] = uvp.cov_array[spw][blpt_inds]
+
+            # fill weighting matrix E
+            if weight_by_cov:
+                # weight by inverse (real) covariance
+                for p in range(uvp.Npols):
+                    E[:, dslice, dslice, p] = np.linalg.pinv(cov_array[spw][:, dslice, dslice, p].real)
```

Comment:
I think here we could add some notations to emphasize the inverse covariance matrix C^{-1} used as weights here, has a shape of (N_cyl, Ncyl). Because we only have the covariance information for bandpowers between different delays on the same baseline-pair, our C here is block diagonal, so C^{-1} is equal to the direct sum of the inverse of each block matrix.

Reply:
sure thing